crates/ra_parser/src/parser.rs

```diff
@@ -228,36 +229,17 @@ impl Marker {
             }
         }
     }
-
-    /// Contract a node `cm` and complete as `cm`'s `kind`.
-    /// `cm` must be a child of `m` to work correctly.
-    /// ```text
-    /// m--A        m--A
-    /// +--cm--B -> +--B
-    /// +--C        C
-    ///
-    /// [m: TOMBSTONE, A, cm: Start(k), B, Finish, C]
-    /// [m: Start(k), A, cm: TOMBSTONE, B, Finish, C]
-    /// ```
-    pub(crate) fn contract_child(mut self, p: &mut Parser, cm: CompletedMarker) -> CompletedMarker {
-        self.bomb.defuse();
-        match p.events[self.pos as usize] {
-            Event::Start { kind: ref mut slot, .. } => *slot = cm.kind(),
-            _ => unreachable!(),
-        };
-        match p.events[cm.0 as usize] {
-            Event::Start { kind: ref mut slot, .. } => *slot = TOMBSTONE,
-            _ => unreachable!(),
-        };
-        CompletedMarker::new(self.pos, cm.kind())
-    }
 }
 
-pub(crate) struct CompletedMarker(u32, SyntaxKind);
+pub(crate) struct CompletedMarker {
+    start_pos: u32,
+    finish_pos: u32,
```

Comment:
I'd rather use an approach which doesn't add this feild to `CompletionMarker`. If we add a field to CM, it means that it can potentially do more, and that makes it harder to understand what it actually does.

Reply:
In other words, being *less* general is a feature :-)

Reply:
Yes, less power is better, if it can be done in that way. However, in the previous method has a bug/confusing behavior https://github.com/rust-analyzer/rust-analyzer/pull/996/commits/ffed132e52964f029a586a3efd4afc879ed39c59#diff-7f6e2fc7cf020c4dc17aa7d4f85ee787R240 here, `C` is not attached to new `m` as expected from method name, but attached to the parent of `m`. In the current usage `C` part is empty and not a problem but this is not good (very ad-hoc). I had to introduce `finish_pos` to fix this behavior and if I introduce `finish_pos` anyway, new API seems cleaner.

Reply:
:ok_hand: , don't have a strong opinion here!